### ISSUE TITLE
Run TI Process After SFTP Transfer Complete

### ISSUE THREAD

**Issue by ToddGaluska** [cubewise-code/tm1py#119]
Marius,

Sorry to post this here if it should be somewhere else. Have you ever initiated a TI Process after an SFTP transfer completed. I am running a python script on Server #1 to transfer a csv file to Server #2 and after it completes I want a TI process to run. Do you have anything like this I couldnt find anything even though I search the tm1py area. 

Thanks

**Comment by MariusWirtz** [comment]
Hi @ecapitaladvisors ,

I have never used this library myself, but for sftp you could perhaps look into [pysftp](https://pysftp.readthedocs.io/en/release_0.2.8/cookbook.html).

Regarding the TI process execution afterwards. This would be fairly simple:

``` python

# copy files through sftp

process_name = "Bedrock.Server.Wait"

with TM1Service(**config['tm1srv01']) as tm1:
    success, state, error_log_file = tm1.processes.execute_with_return(
        process_name=process_name, 
        pWaitSec="1")
```

**Comment by ecapitaladvisors** [comment]
Marius,
I dont know if this will be helpful to other. I created a python script to transfer a csv from one server to another via SFTP and then run a TI process immediately after. We have always been a little constrained by the TM1/PA Chore scheduler that operates off time. This would create no time delay when the file is dropped and the process is run. Let me know what you think?

import paramiko
import configparser
config = configparser.ConfigParser()
config.read(r'C:\Users\xxxx\config.ini')
from TM1py import TM1Service
class SSHConnection(object):
    def __init__(self, host, username, password, port=22):
        self.sftp = None
        self.sftp_open = False
        self.transport = paramiko.Transport((host, port))
        self.transport.connect(username="testuser", password="testuser")
    def _openSFTPConnection(self):
        if not self.sftp_open:
            self.sftp = paramiko.SFTPClient.from_transport(self.transport)
            self.sftp_open = True
    def get(self, remote_path, local_path=None):
        self._openSFTPConnection()
        self.sftp.get(remote_path, local_path)
    def put(self, local_path, remote_path=None):
        self._openSFTPConnection()
        self.sftp.put(local_path, remote_path)
    def close(self):
        if self.sftp_open:
            self.sftp.close()
            self.sftp_open = False
        self.transport.close()

if __name__ == "__main__":
 host = "server where file is"
 username = "testuser"
 pw = "testuser"
 origin = 'C:\Home\Channels.csv'
 dst = 'C:\Tm1_Models\Greatoutdoors\sourcefiles\Channels.csv'
 ssh = SSHConnection(host, username, pw)
 ssh.get(origin, dst)
 with TM1Service(**config['Greatoutdoors']) as tm1_target:
  tm1_target.processes.execute('CreateChannels')
 ssh.close()

**Comment by MariusWirtz** [comment]
Hi,

thanks for the code. This is a common use case I think.
Tried to fix the intentation, but not sure if it's better or worse now :smile:

generally it's good if your wrap your code like this:
``` python
print('Hello World')
```

Strange... I have copied code samples from my IDE (pycharm) before and the indentation was not removed.

Actually the best way to share it though would be to fork the tm1py-samples repository and create a Merge Request that includes your changes.
